### PR TITLE
PP-2670 Get a list of Products by externalServiceId

### PR DESCRIPTION
**Pull request by georgeracu**
- Added resource to GET a list of Product by a
query param externalServiceId;
- Added tests to test endpoint returning 200 and a list of existing
products, 404 when no products are found and 401 when not authenticated;
- Changed tests to allow adding Product and Catalogue in separate
operations;
- Updated LinksDecorator to decorate a list of products;

### CODE REVIEW

File: src/main/java/uk/gov/pay/products/persistence/dao/ProductDao.java [hunk 1]
```diff
@@ -23,4 +27,18 @@ protected ProductDao(Provider<EntityManager> entityManager) {
...
     }
+
+    public List<Product> findByExternalServiceId(String externalServiceId) {
+        String query = "SELECT product FROM ProductEntity product " +
+                "WHERE product.catalogueEntity.externalServiceId = :externalServiceId " +
+                "AND product.status = :status";
+
+        return entityManager.get()
+                .createQuery(query, ProductEntity.class)
+                .setParameter("externalServiceId", externalServiceId)
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/products/persistence/dao/ProductDao.java]: index needed on externalServiceId?

Comment (mrlumbu) [reply src/main/java/uk/gov/pay/products/persistence/dao/ProductDao.java]: and maybe product status?

File: src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java [hunk 2]
```diff
@@ -217,5 +217,60 @@ public void givenANotAuthenticatedRequest_shouldReturn401() throws Exception {
...
+                .then()
+                .statusCode(200);
+
+        response.body("", hasSize(2))
+                .body("[0].external_service_id", matchesPattern(externalServiceId))
+                .body("[0]._links", hasSize(2))
+                .body("[0].description", matchesPattern(product.getDescription()))
+                .body("[0].price", is(product.getPrice().intValue()))
+                .body("[0].name", matchesPattern(product.getName()))
+                .body("[1].name", matchesPattern(product_2.getName()));
```
Comment (mrlumbu) [inline src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java]: why are we not checking all things attributes of product 2?

File: src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java [hunk 3]
```diff
@@ -217,5 +217,60 @@ public void givenANotAuthenticatedRequest_shouldReturn401() throws Exception {
...
+    }
+
+    @Test
+    public void givenNonExistingExternalServiceId_shouldReturn404() throws Exception {
+        givenAuthenticatedSetup()
+                .when()
+                .accept(APPLICATION_JSON)
+                .get(format("/v1/api/products?externalServiceId=%s", randomUuid()))
+                .then()
+                .statusCode(404);
```
Comment (mrlumbu) [inline src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java]: also need to test if an existing service with no active products returns 404?

File: src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java (hunk 2 above)
Comment (georgeracu) [reply src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java]: Given that the above list is generated using a recursive call on a method that was tested earlier to generate the expected output I considered not testing for the second product as the functionality that will generate the values was tested above.

Comment (mrlumbu) [reply src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java]: fair enough 👍

File: src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java (hunk 3 above)
Comment (georgeracu) [reply src/test/java/uk/gov/pay/products/resources/ProductsResourceTest.java]: :1st_place_medal:

File: src/main/resources/migrations/00001_add_catalogues_table.sql [hunk 4]
```diff
@@ -12,3 +12,5 @@ CREATE TABLE catalogues (
 );
 --rollback drop table catalogues;
 
+--changeset uk.gov.pay:add_catalogues_external-service-id_index
+CREATE INDEX external_service_id_idx ON catalogues (external_service_id);
```
Comment (mrlumbu) [inline src/main/resources/migrations/00001_add_catalogues_table.sql]: I don't think this works - liquibase calculates checksums of already executed changes and stores them _somewhere_, so if you modify a changeset that's already been executed it'll complain and fail. I think you need to add a new changeset (`00003_blah`) with these changes. Worth running by Kasun, Raul or someone.

**Pull request merged**